### ISSUE TITLE
intermittent realpath on relative paths (macos 10.15.1, 1.12-dev release)

### ISSUE THREAD

**Issue by pf10** [Clozure/ccl#280]
I have a strange behavior in realpath on 10.15.1 with 1.12-dev release,
sometimes probe-file and other related functions fail to find the file. (I
also have the same issue with master built using macports clang, but I've
reverted to release binaries before attempting to isolate it.)

```
#|
? (probe-file "foo")
#P"/Users/pf/foo"
? (probe-file "foo")
NIL
? (probe-file "foo")
#P"/Users/pf/foo"
? (probe-file "foo")
NIL
? (probe-file "foo")
NIL
|#
```

 I've isolated the issue to realpath intermittently not working for
relative paths, but working reliably for absolute paths. Here's a sample
code,

```
(in-package :ccl)
(defun p (x)
  (without-interrupts
    (without-gcing
      (%stack-block ((resultbuf #$PATH_MAX)
                     (cwd #$PATH_MAX))
        (with-filename-cstrs ((name x))
          (#_getcwd cwd #$PATH_MAX)
          (let ((realpath (foreign-symbol-entry "realpath")))
            (values  
             realpath ;always calling the same realpath
             (%null-ptr-p (%ff-call realpath :address name :address resultbuf :address))
             (%get-errno)
             (%get-cstring cwd))))))))
```

 Running above in a loop with a file "foo" in whatever current directory,
creates a log of mostly fails, with an occasional success, like below:

```
#|
? (loop (pprint (multiple-value-list (p"foo"))))
...
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 NIL 0 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
(140734962563647 T -2 "/Users/pf")
...
|#
```

**Comment by pf10** [comment]
so i wrapped the realpath in

```
#include<stdlib.h>
#include<stdio.h>
char*foo(const char *filename, char *resolved){
	printf("-> %s %s\n", filename, resolved);
	char *x=realpath(filename, resolved);
	printf("<- %s\n", x);
	return x;
}
```

and patched `p` above to call `foo` instead, but with that piece of scaffolding i can't reproduce the issue

**Comment by xrme** [comment]
I've seen symptoms like this too. I wonder what realpath's problem is.

**Comment by pf10** [comment]
@xrme  is it possible somehow related to there being two versions of `realpath` in the system library?

```
230	#if (__DARWIN_UNIX03 && !defined(_POSIX_C_SOURCE)) || defined(_DARWIN_C_SOURCE) || defined(_DARWIN_BETTER_REALPATH)
231	char	*realpath(const char * __restrict, char * __restrict) __DARWIN_EXTSN(realpath);
232	#else /* (!__DARWIN_UNIX03 || _POSIX_C_SOURCE) && !_DARWIN_C_SOURCE && !_DARWIN_BETTER_REALPATH */
233	char	*realpath(const char * __restrict, char * __restrict) __DARWIN_ALIAS(realpath);
234	#endif /* (__DARWIN_UNIX03 && _POSIX_C_SOURCE) || _DARWIN_C_SOURCE || _DARWIN_BETTER_REALPATH */
```

 i don't understand enough of the ffi mechanism yet to debug it further..

**Comment by xrme** [comment]
On my system, I see three realpath variants:

	realpath             
	realpath$DARWIN_EXTSN
	realpath_np          

We are using the plain realpath.  For some reason, plain realpath seems to be flaky on 10.15.  If I use `(external-call "realpath$DARWIN_EXTSN" :address name :address resultbuf :address)` in place of `(#_realpath name resultbuf)`, I don't see the flakiness.  I believe a vanilla C program would end up using `realpath$DARWIN_EXTSN` as well.

I don't like exposing that internal symbol name, though.  I think the right thing is to add a `lisp_realpath` kernel import, like I did with `lisp_lstat` in 479244ee.  Then, we will benefit from the C include file tricks that the operating system suppliers provide.  I have a patch locally that does just that, and I'll be committing it shortly.

It would be nice if dlsym would pick up the right variant of realpath depending on how the lisp kernel was linked (via -mmacosx-version or something), but maybe shared libraries can't work that way for binary compatibility reasons?

**Comment by xrme** [comment]
This is fixed in 1a304fd9.

You must rebuild the lisp kernel before doing `(rebuild-ccl :clean t)`.

**Comment by eschaton** [comment]
You’re correct, shared libraries can’t really work that way for binary-compatibility reasons.
